compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -205,6 +210,157 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            }
+        }
+
+        let hir::ExprKind::Path(hir::QPath::Resolved(None, p)) = expr.kind else { return false; };
+        let [hir::PathSegment { ident, args: None, .. }] = p.segments else { return false; };
+        let hir::def::Res::Local(hir_id) = p.res else { return false; };
+        let Some(node) = map.find(hir_id) else { return false; };
+        let hir::Node::Pat(pat) = node else { return false; };
+        let parent = map.get_parent_node(pat.hir_id);
+        let Some(hir::Node::Local(hir::Local {
+            ty: None,
+            init: Some(init),
+            ..
+        })) = map.find(parent) else { return false; };
+
+        let ty = self.node_ty(init.hir_id);
+        if ty.is_closure() || init.span.overlaps(expr.span) {
+            return false;
+        }
+        let mut span_labels = vec![(
+            init.span,
+            with_forced_trimmed_paths!(format!(
+                "here the type of `{ident}` is inferred to be `{ty}`",
+            )),
+        )];
+
+        // Locate all the usages of the relevant binding.
+        struct FindExprs<'hir> {
+            hir_id: hir::HirId,
+            uses: Vec<&'hir hir::Expr<'hir>>,
+        }
+        impl<'v> Visitor<'v> for FindExprs<'v> {
+            fn visit_expr(&mut self, ex: &'v hir::Expr<'v>) {
+                if let hir::ExprKind::Path(hir::QPath::Resolved(None, path)) = ex.kind
+                    && let hir::def::Res::Local(hir_id) = path.res
+                    && hir_id == self.hir_id
+                {
+                    self.uses.push(ex);
+                }
+                hir::intravisit::walk_expr(self, ex);
+            }
+        }
+
+        let mut expr_finder = FindExprs { hir_id, uses: vec![] };
+        let id = map.get_parent_item(hir_id);
+        let hir_id: hir::HirId = id.into();
+
+        if let Some(node) = map.find(hir_id) && let Some(body_id) = node.body_id() {
+            let body = map.body(body_id);
+            expr_finder.visit_expr(body.value);
+            let mut eraser = TypeEraser { tcx };
+            let mut prev = eraser.fold_ty(ty);
+
+            for ex in expr_finder.uses {
+                if ex.span.overlaps(expr.span) { break; }
+                let parent = map.get_parent_node(ex.hir_id);
+                if let Some(hir::Node::Expr(expr))
+                | Some(hir::Node::Stmt(hir::Stmt {
+                    kind: hir::StmtKind::Expr(expr) | hir::StmtKind::Semi(expr),
+                    ..
+                })) = &map.find(parent)
+                    && let hir::ExprKind::MethodCall(s, rcvr, args, span) = expr.kind
+                    && rcvr.hir_id == ex.hir_id
+                {
+                    let ty = if let Ok(m) = self.lookup_method(ty, s, span, expr, rcvr, args) {
+                        // We get the self type from `lookup_method` because the `rcvr` node
+                        // type will not have had any adjustments from the fn arguments.
```

Comment:
What about `node_ty_adjusted`? Doesn't that show us the adjusted type of the rcvr expr?

If that doesn't work, then I'd really prefer if you avoided this extra lookup with something by using the already computed method def-id for a hir expr by using a combination of these two methods:
1. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def
2. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.node_substs

Reply:
`expr_ty_adjusted` gives us the following:

```rust
% rustc +dev f57.rs -Zverbose
error[E0308]: mismatched types
  --> f57.rs:12:9
   |
6  |     let mut foo = vec![];
   |                   ------ here the type of `foo` is inferred to be `Vec<_#16t>`
...
9  |         foo.push(i);
   |         --- here the type of `foo` is inferred to be `&ReErased mut Vec<_#0t>`
10 |     }
11 |     baz(&foo);
   |          --- here the type of `foo` is inferred to be `Vec<&ReErased _#0i>`
12 |     bar(foo);
   |     --- ^^^ expected `i32`, found `&_#0i`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `Vec<i32>`
              found struct `Vec<&'_#14r _#0i>`
```

Notice that `_#0t` is the inference variable on `foo`, so `i` hasn't affected it yet. I also still need to remove the self receiver borrow.

Reply:
Interestingly, with the following:

```rust
                    && let Some(def_id) = self.typeck_results.borrow().type_dependent_def_id(expr.hir_id)
                    && let Some(substs) = self.typeck_results.borrow().node_substs_opt(expr.hir_id)
                {
                    let sig = self.tcx.fn_sig(def_id);
                    let method_ty = tcx.mk_fn_def(def_id, substs);
                    err.note(&format!("\n{:?}\n{:?}", sig, method_ty.kind()));
```

the output is

```
   = note:
              Binder(([&ReLateBound(DebruijnIndex(0), BoundRegion { var: 0, kind: BrNamed(DefId(5:8965 ~ alloc[b102]::vec::{impl#1}::push::'_), '_) }) mut std::vec::Vec<T, A>, T]; c_variadic: false)->(), [Region(BrNamed(DefId(5:8965 ~ alloc[b102]::vec::{impl#1}::push::'_), '_))])
              FnDef(DefId(5:6488 ~ alloc[b102]::vec::{impl#1}::push), [_#37t, _#38t])
```

which looks to me to have only inference variables, because the `node_substs` for `expr.hir_id` do not have the changes introduced by `push(i)` yet.

I also wanted to reuse as much of the existing machinery as possible, as I don't want it to diverge after a refactor. I'm looking if calling `prove_op` directly is feasible from here instead, just to see if the type is in `pick` already.

Reply:
Do let me know what you think of the current approach. It's not quite what you asked, but I think it is robust.